Automattic/woocommerce-payments
# Add: proxy endpoints for terminal locations API

Fixes #2334 

#### Changes proposed in this Pull Request
- Exposes the `DELETE /terminal/locations/{{location_id}}` endpoint which proxies the delete terminal endpoint to server introduced in [#1194](https://github.com/Automattic/woocommerce-payments-server/pull/1194)
- Exposes the  `POST /terminal/location/{{location_id}}` endpoint which proxies the update terminal endpoint to server introduced in [#1193](https://github.com/Automattic/woocommerce-payments-server/pull/1193)
- Exposes the `GET /terminal/locations/{{location_id}}` endpoint which proxies the get terminal request to server introduced in [#1191](https://github.com/Automattic/woocommerce-payments-server/pull/1191)
- Expose the `POST /terminal/locations` endpoint which proxies the create terminal request to server introduced in [#1145](https://github.com/Automattic/woocommerce-payments-server/pull/1145)
- Expose the `GET /terminal/locations` endpoint which proxies the get all terminal locations request to server introduced in [#1145](https://github.com/Automattic/woocommerce-payments-server/pull/1145) 

#### Testing instructions
- DELETE endpoint
  - Get a store id using the `GET /terminal/locations/store` endpoint.
  - Make a request to the newly introduced `DELETE /terminal/locations/{{location_id}}` to delete it.
  - Verify the cache entry for the key `wcpay_store_terminal_locations` was deleted.
  - Verify the endpoint is idempotent and returns the same response for deleted records.
  - ![Screenshot 2021-09-30 at 1 14 36 PM](https://user-images.githubusercontent.com/68396823/135409539-81a2d51a-d367-4cf2-b08e-4fa0d99bb09b.png)
 - UPDATE endpoint
   - Get a store id using the `GET /terminal/locations/store` endpoint.
   - Make a request to the newly introduced `POST /terminal/location/{{location_id}}` to update it by passing `display_name` or `address` or both in the request body.
   - Verify the response has the updated fields set based on the request body.
   - Verify the cache entry for the key `wcpay_store_terminal_locations` was deleted.
   - Verify that get store endpoint returns a new terminal location id now since cache lookup fails.
   - ![Screenshot 2021-10-01 at 3 30 45 PM](https://user-images.githubusercontent.com/68396823/135602053-75eca265-cd75-48c0-b2cb-726c8af9c7f6.png)
 - CREATE endpoint
   -  Verify the request treats `display_name` and `address` as required fields and fails with validation errors in their absence.
   - Verify that valid parameters create a new terminal location and return the success response.
   - Verify that the newly created location is added to the cache entry.
   - ![Screenshot 2021-10-01 at 6 18 44 PM](https://user-images.githubusercontent.com/68396823/135622070-d07743e2-6821-4b92-bdbc-cce99bf1ca28.png)
- GET single location endpoint
  - Verify that on passing a valid location id, the location details are fetched.
  - Verify that invalid location ids lead to a not found error.
  - ![Screenshot 2021-10-01 at 6 19 13 PM](https://user-images.githubusercontent.com/68396823/135622127-e3bfc910-9dbc-4176-9807-cc700aa493c2.png)
- GET all locations endpoint
  - Verify that correctly formatted location response is returned.
  - Verify that cache is used for returning the location response if it exists.
  - Verify that cache is set with fetched location data in case of a cache miss.
  -  ![Screenshot 2021-10-05 at 6 10 38 PM](https://user-images.githubusercontent.com/68396823/136024252-ac120ff4-0c20-4cab-981a-b8d4c51f9a2c.png)


<b>Note</b>: The update and retrieve endpoints are still not merged at the server yet so testing them would require switching server branches for now.

-------------------
